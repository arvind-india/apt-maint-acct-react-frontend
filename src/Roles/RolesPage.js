import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table, Alert, UncontrolledAlert } from 'reactstrap'
import { Router, Route } from 'react-router-dom'

import {
  MdAdd,
  MdVisibility,
  MdDelete
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { roleActions as actions, alertActions } from '../_actions'
import { RoleDetailsPage as detailsPage } from './RoleDetailsPage'

let url = '/roles'

class RolesPage extends React.Component {

  constructor(props) {
    super(props)
    this.handleDeleteModel = this.handleDeleteModel.bind(this)
  }

  componentDidMount() {
    this.props.dispatch(actions.getAll())
  }
  handleDeleteModel(id) {
    console.log('Deleting Role with id: ', id)
    //return (e) => this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.getAll()) // get list after deletion of a model
  }
  showList(models){
    return <Table>
      <thead>
        <tr>
          <th>#</th>
          <th>Role Name</th>
          <th>Description</th>
          <th>Inherits</th>
          <th>Actions <Link to={`${url}/0`} title="Add"><MdAdd/></Link></th>
        </tr>
      </thead>
      <tbody>
        {models.items.map((model, index) =>
          <tr key={model.id}>
            <th scope="row">{index+1}</th>
            <td>{model.name}</td>
            <td>{model.description}</td>
            <td>{model.inherits}</td>
            <td>
              <Link
                to={`${url}/${model.id}`}
                title="View or Edit"
              ><MdVisibility/></Link>
              <Button
                color="link"
                title="Delete"
                onClick={() => this.handleDeleteModel(model.id)}
              ><MdDelete color="red"/></Button>
            </td>
          </tr>)}
      </tbody>
    </Table>
  }

  render() {
    console.log('Props in RolesPage: ', this.props)
    const { user, roles, alert } = this.props
    let models = roles
    return (
      <div>
        <h3>Roles List</h3>
        {alert.message && <FlashMessage text={alert.message} delay={5000}/>}
        {models.loading && <em>Loading models...}</em>}
        {models.error && <span className="text-danger">ERROR: {models.error}</span>}
        {models.items && this.showList(models) }
        <Router history={history}>
          <div>
            <PrivateRoute path={`${url}/:id`} component={detailsPage} />
          </div>
        </Router>
      </div>
    )
  }
}

function mapStateToProps(state) {
  const { roles, authentication, alert } = state
  const { user } = authentication
  return {
    user,
    roles,
    alert
  }
}

const connectedRolesPage = connect(mapStateToProps)(RolesPage)
export { connectedRolesPage as RolesPage }
