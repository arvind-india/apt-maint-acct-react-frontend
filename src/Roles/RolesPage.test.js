// RolesPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { Roles } from './RolesPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  roles: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        name: 'admin',
        description:'administrator with all access',
        inherits: ''
      },
      {
        id: 2,
        name: 'supervisor',
        description:'supervisor with limited access compared to administrator',
        inherits: ''
      },
      {
        id: 3,
        name: 'owner',
        description:'Owner can access own records for editing purpose',
        inherits: 'tenant'
      }
    ]
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAll: jest.fn(),
  delete: jest.fn()
}

const component = shallow(<Roles {...mockProps} />)

describe('RolesPage params test', () => {
  it('should render RolesPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})

describe('RolesPage rows test', () => {
  let numberOfMockRoles = mockProps.roles.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockRoles+' rows', () => {
    expect(rows.length).toEqual(numberOfMockRoles)
  })

})
