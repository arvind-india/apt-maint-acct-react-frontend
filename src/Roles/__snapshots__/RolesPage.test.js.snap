// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolesPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Roles List
    </h3>
    <FlashMessage
        delay={4000}
        text="testingAlert"
    />
    <Table
        responsiveTag="div"
        tag="table"
    >
        <thead>
            <tr>
                <th>
                    #
                </th>
                <th>
                    Role Name
                </th>
                <th>
                    Description
                </th>
                <th>
                    Inherits
                </th>
                <th>
                    Actions 
                    <Link
                        replace={false}
                        title="Add"
                        to={
                            Object {
                                "pathname": "/roles/0",
                                "state": Object {
                                  "model": Object {
                                    "description": "",
                                    "id": 0,
                                    "inherits": "",
                                    "name": "",
                                  },
                                },
                              }
                        }
                    >
                        <MdAdd />
                    </Link>
                </th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <th
                    scope="row"
                >
                    1
                </th>
                <td>
                    admin
                </td>
                <td>
                    administrator with all access
                </td>
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/roles/1",
                                "state": Object {
                                  "model": Object {
                                    "description": "administrator with all access",
                                    "id": 1,
                                    "inherits": "",
                                    "name": "admin",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    2
                </th>
                <td>
                    supervisor
                </td>
                <td>
                    supervisor with limited access compared to administrator
                </td>
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/roles/2",
                                "state": Object {
                                  "model": Object {
                                    "description": "supervisor with limited access compared to administrator",
                                    "id": 2,
                                    "inherits": "",
                                    "name": "supervisor",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    3
                </th>
                <td>
                    owner
                </td>
                <td>
                    Owner can access own records for editing purpose
                </td>
                <td>
                    tenant
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/roles/3",
                                "state": Object {
                                  "model": Object {
                                    "description": "Owner can access own records for editing purpose",
                                    "id": 3,
                                    "inherits": "tenant",
                                    "name": "owner",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
        </tbody>
    </Table>
    <Router
        history={Object {}}
    >
        <div>
            <PrivateRoute
                component={[Function]}
                path="/roles/:id"
            />
        </div>
    </Router>
</div>,
]
`;
